orc8r/cloud/go/service/middleware/unary/identity_decorator.go

```diff
@@ -112,70 +112,97 @@ func SetIdentityFromContext(ctx context.Context, _ interface{}, info *grpc.Unary
+		if !allow {
 			err = status.Error(codes.Unauthenticated, "Multiple CSNs present")
-		} else {
-			// One CSN is found, find Identity associated with it
-			var gwIdentity *protos.Identity
-			var certExpTime int64
-			// Check if SN is the reserved value used for all inter-orc8r calls
-			if snlist[0] == unarylib.ORC8R_CLIENT_CERT_VALUE {
-				return newCtx, newReq, resp, nil
-			}
-			gwIdentity, certExpTime, err = findGatewayIdentity(ctx, snlist[0], ctxMetadata)
-			if err == nil {
-				// If a valid GW Identity is found, add it into CTX for use
-				// by the callee
-				newCtx = gwIdentity.NewContextWithIdentity(protos.NewContextWithCertExpiration(ctx, certExpTime))
-				return newCtx, newReq, resp, err
-			}
 		}
-	} else {
-		// No CSNs found, check there is also no Certificate Common Names
-		if _, ok = ctxMetadata[CLIENT_CERT_CN_KEY]; ok {
-			// CN header is present while SN header is missing - possible
-			// security hack, either both or neither of the headers should be
-			// set
-			glog.Infof("CCN is present without SCN in metadata: %+v", ctxMetadata)
+	} else if snok {
+		// 5.
+		// One CSN is found, find Identity associated with it
+		newCtx, err = findIdentityFromCertificateSerialNumber(snlist, ctx, ctxMetadata)
+		if allow {
+			err = nil
+		}
+	} else if _, ok = ctxMetadata[CLIENT_CERT_CN_KEY]; ok {
+		// 4.
+		// No CSNs found, check there is also no Certificate Common Names.
+		// CN header is present while SN header is missing - possible
+		// security hack, either both or neither of the headers should be
+		// set
+		glog.Infof("CCN is present without SCN in metadata: %+v", ctxMetadata)
+		if !allow {
 			err = status.Error(codes.Unauthenticated, "Inconsistent Request Signature")
 		}
+	} else if !allow {
+		// No CN header and no SN header are present: only allowlisted
+		// calls and internal calls are allowed
+		err = isLocalClientCall(ctx, info)
 	}
+	return newCtx, newReq, resp, err
+}
 
+// isAllowlistedCall checks if the call is from an allowlisted method
+func isAllowlistedCall(info *grpc.UnaryServerInfo) bool {
 	if info != nil {
 		// Check if the call is for a allowlisted method - anything is allowed
 		// do this check past possible identity decoration to still allow to add
 		// valid identity even to allowlisted requests
-		if _, ok := identityDecoratorBypassList[info.FullMethod]; ok {
-			// Bypass method (Bootstrapper & Co.), shortcut...
-			return newCtx, newReq, resp, nil
-		}
+		_, ok := identityDecoratorBypassList[info.FullMethod]
+
+		// Bypass method (Bootstrapper & Co.), shortcut...
+		return ok
 	}
+	return false
+}
 
-	// Only allow local clients if there is no previous errors (snok == false)
+// findIdentityFromCertificateSerialNumber will try to find the identity associated
+// with the one caller's Client Certificate Serial Number found in CTX
+func findIdentityFromCertificateSerialNumber(snlist []string, ctx context.Context, ctxMetadata metadata.MD) (newCtx context.Context, err error) {
```

Comment:
maybe makes sense to just pass a single string snlist[0] if the function is only using index 0. It'd also remove from the caller a burden of enforcing explicit & undocumented condition of passing of the lis list of len >= 1.

Reply:
Good point, thanks!